File: 2015/16/javascript/main.mjs
import fs from "node:fs";
import readline from "node:readline";

// --- Constants ---
const INPUT_FILE = "../input/input.txt";
const REGEX = /([a-z]+): (\d+)/g;

// --- MFCSAM Output ---
const MAP = {
	children: 3,
	cats: 7,
	samoyeds: 2,
	pomeranians: 3,
	akitas: 0,
	vizslas: 0,
	goldfish: 5,
	trees: 3,
	cars: 2,
	perfumes: 1
};

// --- Functions ---
function find1(line) {
	for (const cap of line.matchAll(REGEX)) {
		const key = cap[1];
		const value = Number(cap[2]);
		if (MAP.hasOwnProperty(key) && MAP[key] !== value) return false;
	}
	return true;
}

function find2(line) {
	for (const cap of line.matchAll(REGEX)) {
		const key = cap[1];
		const value = Number(cap[2]);
		if (MAP.hasOwnProperty(key)) {
			if (key === "cats" || key === "trees") {
				if (MAP[key] >= value) return false;
			} else if (key === "pomeranians" || key === "goldfish") {
				if (MAP[key] <= value) return false;
			} else {
				if (MAP[key] !== value) return false;
			}
		}
	}
	return true;
}

// --- Variables ---
let sue1 = null;
let sue2 = null;

// --- Read and parse the input file ---
const stream = fs.createReadStream(INPUT_FILE);
const reader = readline.createInterface({ input: stream, crlDelay: Infinity });
reader.on("line", line => {
	// --- Puzzle 1 ---
	if (sue1 === null && find1(line)) {
		sue1 = line.split(":")[0];
	}

	// --- Puzzle 2 ---
	if (sue2 === null && find2(line)) {
		sue2 = line.split(":")[0];
	}
});
await new Promise(res => reader.once("close", res));

// --- Results ---
console.log("1. Number of the Sue that got you the gift: ", sue1);
console.log("2. Number of the Sue that got you the gift: ", sue2);

File: 2015/16/python/main.py
import re

# --- Constants --
INPUT_FILE = "../input/input.txt"
REGEX = r"([a-z]+): (\d+)"

# --- MFCSAM Output ---
MAP = {
    'children': 3,
    'cats': 7,
    'samoyeds': 2,
    'pomeranians': 3,
    'akitas': 0,
    'vizslas': 0,
    'goldfish': 5,
    'trees': 3,
    'cars': 2,
    'perfumes': 1
}

# --- Functions ---
def find1(line):
    for cap in re.findall(REGEX, line):
        key = cap[0]
        value = int(cap[1])
        if key in MAP and MAP[key] != value: return False
    return True

def find2(line):
    for cap in re.findall(REGEX, line):
        key = cap[0]
        value = int(cap[1])
        if key in MAP:
            if key == "cats" or key == "trees":
                if MAP[key] >= value: return False
            elif key == "pomeranians" or key == "goldfish":
                if MAP[key] <= value: return False
            else:
                if MAP[key] != value: return False
    return True

# --- Variables ---
sue1 = ""
sue2 = ""

# --- Read and Parse the input file --
file = open(INPUT_FILE, "r")
while True:
    line = file.readline()
    if not line: break
    line = line.strip()

    if sue1 == "" and find1(line): sue1 = line.split(":")[0]
    if sue2 == "" and find2(line): sue2 = line.split(":")[0]
file.close()

# --- Results ---
print(f"1. Number of the Sue that got you the gift: {sue1}")
print(f"2. Number of the Sue that got you the gift: {sue2}")
